src/tracing/mod.rs

```diff
@@ -0,0 +1,137 @@
+        let maybe_block = match block_id {
+            BlockId::Hash(hash) => self.eth_provider.block_by_hash(hash.block_hash, true).await?,
+            BlockId::Number(number) => self.eth_provider.block_by_number(number, true).await?,
+        };
+
+        if maybe_block.is_none() {
+            return Ok(None);
+        }
+
+        let block = maybe_block.unwrap();
+        if block.header.hash.unwrap_or_default().is_zero() {
+            return Err(EthApiError::UnknownBlock);
+        }
+
+        // DB should use the state of the parent block
+        let parent_block_hash = block.header.parent_hash;
+        let db = database::EthDatabaseSnapshot::new(self.eth_provider.clone(), BlockId::Hash(parent_block_hash.into()));
+
+        let env = Box::new(self.setup_env(block.header.clone()).await);
+        let ctx = reth_revm::Context::new(reth_revm::EvmContext::new_with_env(db, env.clone()), self.inspector.clone());
+
+        let env = EnvWithHandlerCfg::new(env, HandlerCfg::new(SpecId::CANCUN));
+        let mut handler = reth_revm::Handler::new(env.handler_cfg);
+        handler.append_handler_register_plain(reth_revm::inspector_handle_register);
+        let mut evm = reth_revm::Evm::new(ctx, handler);
+
+        let traces = tokio::task::block_in_place(move || {
+            let transactions = match &block.transactions {
+                BlockTransactions::Full(transactions) => transactions,
+                _ => return Err(TransactionError::ExpectedFullTransactions.into()),
+            };
+            let mut transactions = transactions.iter().peekable();
+            let mut traces = Vec::with_capacity(block.transactions.len());
+
+            while let Some(tx) = transactions.next() {
+                // Convert the transaction to an ec recovered transaction and update the evm env with it
+                let tx_ec_recovered = rpc_to_ec_recovered_transaction(tx.clone())?;
+                let tx_env = tx_env_with_recovered(&tx_ec_recovered);
+                evm = evm.modify().modify_tx_env(|env| *env = tx_env).build();
+
+                // Transact the transaction
+                let result = evm.transact().map_err(|err| TransactionError::Tracing(err.into()))?;
+
+                // Convert the traces to a parity trace and accumulate them
+                let parity_builder = evm.context.external.clone().into_parity_builder();
+
+                // Extract the base fee from the transaction based on the transaction type
+                let base_fee = match tx.transaction_type.map(|typ| typ.to::<u8>()) {
+                    Some(0) | Some(1) => tx.gas_price,
+                    Some(2) => tx
+                        .max_fee_per_gas
+                        .map(|fee| fee.saturating_sub(tx.max_priority_fee_per_gas.unwrap_or_default())),
+                    _ => return Err(EthereumDataFormatError::TransactionConversionError.into()),
```

Comment:
```suggestion
                let base_fee = match tx.transaction_type.map(|typ| typ.into()) {
                    Some(TxType::Legacy) | Some(TxType::Eip2930) => tx.gas_price,
                    Some(TxType::Eip1559) => tx
                        .max_fee_per_gas
                        .map(|fee| fee.saturating_sub(tx.max_priority_fee_per_gas.unwrap_or_default())),
                    _ => return Err(EthereumDataFormatError::TransactionConversionError.into()),
```

Reply:
Unfortunately `TryFrom<U8>` isn't implemented, which means I need to convert to `u8` first and then use `try_into`.

Reply:
Maybe it's safer no? I mean, I guess the indexes won't change and it will stay stable, but if one day there is a change it's the kind of bug that takes hours to find right? But since there is no big risk, if you prefer we can leave it like that I think.

Reply:
made the changes, just meant that I couldn't use the exact suggestions you made but using TxType is safer indeed :)